src/state.rs

```diff
@@ -103,13 +110,45 @@ impl State {
         } else {
             self.board.move_piece(from, to)
         };
+        let next_castling = self.castling.after_move(self.player, from);
         let next_state = State {
             board: next_board,
             player: self.player.other(),
             en_passant: self.en_passant_pos(from, to),
+            castling: next_castling,
         };
         Move {
-            index: (from, to),
+            action: Action::Simple { from, to },
+            next: next_state,
+        }
+    }
+
+    fn can_castle_without_check(&self, kingside: bool) -> bool {
+        let king_pos = self.board.get_king_pos(self.player);
+        let (pos_1, pos_2) = Castling::king_tracks(self.player, kingside);
+
+        !self.in_check()
+            && !self.move_puts_current_player_in_check(king_pos, pos_1)
```

Comment:
Is this checking whether squares are under attack by making a hypothetical illegal move? If so, that seems a little unintuitive to me.

Reply:
yeah. it is weird, but i'm not actually sure what a better way would be. "if your king were here, would it be in check?" seems like the question you have to ask when you castle?